components/snap_recovery/src/region_meta_collector.rs

```diff
@@ -0,0 +1,229 @@
+}
+
+/// `RegionMetaCollector` is the collector that collector all region meta
+pub struct RegionMetaCollector<ER: RaftEngine> {
+    /// The engine we are working on
+    engines: Engines<RocksEngine, ER>,
+    /// region meta report to br
+    tx: UnboundedSender<RegionMeta>,
+    /// Current working workers
+    worker_handle: RefCell<Option<JoinHandle<()>>>,
+}
+
+#[allow(dead_code)]
+impl<ER: RaftEngine> RegionMetaCollector<ER> {
+    pub fn new(engines: Engines<RocksEngine, ER>, tx: UnboundedSender<RegionMeta>) -> Self {
+        RegionMetaCollector {
+            engines,
+            tx,
+            worker_handle: RefCell::new(None),
+        }
+    }
+    /// Start a collector and region meta report.
+    pub fn start_report(&self) {
+        let worker = CollectWorker::new(self.engines.clone(), self.tx.clone());
+        let props = tikv_util::thread_group::current_properties();
+        *self.worker_handle.borrow_mut() = Some(
+            std::thread::Builder::new()
+                .name("collector_region_meta".to_string())
+                .spawn_wrapper(move || {
+                    tikv_util::thread_group::set_properties(props);
+                    tikv_alloc::add_thread_memory_accessor();
+
+                    worker
+                        .collect_report()
+                        .expect("collect region meta and report to br failure.");
+
+                    tikv_alloc::remove_thread_memory_accessor();
+                })
+                .expect("failed to spawn collector_region_meta thread"),
+        );
+    }
+
+    // join and wait until the thread exit
+    pub fn wait(&self) {
+        if let Err(e) = self.worker_handle.take().unwrap().join() {
+            error!("failed to join thread: {:?}", e);
+        }
+    }
+}
+
+struct CollectWorker<ER: RaftEngine> {
+    /// The engine we are working on
+    engines: Engines<RocksEngine, ER>,
+    tx: UnboundedSender<RegionMeta>,
+}
+
+impl<ER: RaftEngine> CollectWorker<ER> {
+    pub fn new(engines: Engines<RocksEngine, ER>, tx: UnboundedSender<RegionMeta>) -> Self {
+        CollectWorker { engines, tx }
+    }
+
+    fn get_local_region(&self, region_id: u64) -> Result<LocalRegion> {
+        let raft_state = box_try!(self.engines.raft.get_raft_state(region_id));
+
+        let apply_state_key = keys::apply_state_key(region_id);
+        let apply_state = box_try!(
+            self.engines
+                .kv
+                .get_msg_cf::<RaftApplyState>(CF_RAFT, &apply_state_key)
+        );
+
+        let region_state_key = keys::region_state_key(region_id);
+        let region_state = box_try!(
+            self.engines
+                .kv
+                .get_msg_cf::<RegionLocalState>(CF_RAFT, &region_state_key)
+        );
+
+        match (raft_state, apply_state, region_state) {
+            (None, None, None) => Err(Error::NotFound(format!("info for region {}", region_id))),
+            (raft_state, apply_state, region_state) => {
```

Comment:
Perhaps don't make the fields of `LocalRegion` become `Option`s, just report an error here when we cannot find the local region status.